src/torch_ucc.cpp

```diff
@@ -184,9 +184,12 @@ c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupUCC::WorkUCC::getFuture() {
 }
 
 void ProcessGroupUCC::ProgressEntry::finalize(std::exception_ptr eptr) {
-  ucc_status_t status;
-  status = request_->status;
-  comm_->free_request(request_);
+  ucc_status_t status = UCC_OK;
+
+  if (request_ != nullptr) {
+    status = request_->status;
+    comm_->free_request(request_);
+  }
   status_ = status;
   eptr_ = eptr;
```

Comment:
Current timeout test is failing even if UCC detects and reports timeout (https://github.com/facebookresearch/torch_ucc/pull/38/checks?check_run_id=3874227623). I could reproduce it locally (~40% of runs don't catch timeout correctly).

I suspect the main thread detects status (`entry_->status_ <= 0`) before progress thread sets the exception. should we try setting the `eptr_` first? which works for me locally.

```suggestion
  ucc_status_t status = UCC_OK;
  eptr_ = eptr;
    
  if (request_ != nullptr) {
    status = request_->status;
    comm_->free_request(request_);
  }
  status_ = status;
```

Reply:
agree, but perhaps we shouldn't change status at all in case of exception, otherwise there is a race condition in isCompleted() and in isSuccess(), wdyt?

Reply:
done